### ISSUE TITLE
Add National Library of Ukraine to Trusted Book Providers

### ISSUE THREAD

**Issue by cdrini** [internetarchive/openlibrary#7045]
> By the way, this can also be used for Trusted book providers as this library also provides online version of books. For example, let's take book http://irbis-nbuv.gov.ua/ulib/item/UKR0001691 from this page you can click on "Переглянути документ" button, "View the document" if your language is set to English, which will lead you to scan of the book. 
> ![image](https://user-images.githubusercontent.com/58372007/193594359-2f4aa72b-92d4-427a-83cf-435c3828eddd.png)
> 
> Here is the link to online version of this book http://irbis-nbuv.gov.ua/cgi-bin/ua/elib.exe?Z21ID=&I21DBN=UKRLIB&P21DBN=UKRLIB&S21STN=1&S21REF=10&S21FMT=online_book&C21COM=S&S21CNR=20&S21P01=0&S21P02=0&S21P03=FF=&S21STR=ukr0001435

Steps:

- [ ] Get a data dump of records we can import into Open Library with `nbuv` IDs
     - Books in NBUV: ????
     - Books in Open Library with NBUV IDs: [0](http://openlibrary.org/query.json?type=/type/edition&identifiers.nbuv~=*&limit=5)
- [ ] Import all the books into open library
- [ ] Extend `BookProvider` for nbuv to display Read/other buttons
    - Determine how to compute the ID?
    - Determine which records are readable somehow?
- [ ] IDEAL: Set up automatic import pipeline

### Stakeholders
@bicolino34 @el4ctr0n

**Comment by bicolino34** [comment]
It would be quite valuable source of information. There is a page for books that were made available on the site in 2022 and as of 4th October there are 1374 of them.

**Comment by bicolino34** [comment]
@cdrini I changed amount of books to unknown. As the number I sent was only books added in 2022. As far as I know, the online version of library exists since early 2010-s

**Comment by cdrini** [comment]
It's down for me now :( I've escalated to Mark Graham, the director of Wayback Machine; will update once I get an update.

**Comment by cdrini** [comment]
@bicolino34 Do you think you could contact them? Their book scans appear to be stored pretty simply in a file system, and could be backed up with a basically a tar gzip of the directory containing all the PDFs. Mek, Jim and I will discuss it tomorrow and see if we can think of what specifically to ask them. It might be helpful simply to schedule a call with them.

**Comment by cdrini** [comment]
Mark forwarded this to https://www.sucho.org/

**Comment by storytracer** [comment]
Hi everyone, my name is Seb and I'm one of the SUCHO co-founders!

It's important to know that the servers of NBUV have been inherently unstable since before the invasion. We crawled the entire catalogue back in spring and there were lots of timeouts back then already. We are directly in touch with the library and know that their servers have been plagued with SPAM and other issues. So even though it might seem the website is down, it probably isn't and will come back up again. Nevertheless it is important to crawl and back up everything of course.

So fortunately some of our volunteers have managed to save all digitized PDFs from the digital collections of the Vernadyski National Library (NBUV) in spring already (there are several national libraries in Ukraine, the other big one in Kyiv is the [Yaroslav the Wise National Library of Ukraine](https://nlu.org.ua/)). The dump contains 8803 PDF files and is approximately 300GB large. The dump contains all PDFs stored under the path `http://irbis-nbuv.gov.ua/E_LIB/PDF/` which contains PDFs numbered from `UKR0000001.pdf` through to `UKR0008914.pdf` (the total number of PDFs does not match the last numeric ID, because there are random gaps in the numbered files).

The dump is a blessing and a curse at once. Because the IRBIS catalogue system, which the NBUV uses, is just an awful piece of software (unfortunately very widely used in Eastern Europe), there is no consistency in terms of identifiers or permalinks. There are two main digital collections: [Electronic library "Ukrainica", a.k.a ULIB or UKRLIB](http://irbis-nbuv.gov.ua/cgi-bin/ua/elib.exe?C21COM=F&I21DBN=UKRLIB&P21DBN=UKRLIB) and the [Digital Library of Historical and Cultural Heritage a.k.a ELIB or DLIB](http://irbis-nbuv.gov.ua/cgi-bin/irbis_ir/cgiirbis_64.exe?C21COM=F&I21DBN=NAV&P21DBN=ELIB). Neither of the collections can be access by a permanent URL, but are only reachable through a crazy combination of params to the various `exe` files.

Here's the worst thing: the ULIB and ELIB collections have separate incremental numeric identifiers, which each start at 0 and which **do not correspond with the PDF file names**. As a matter of fact multiple PDFs can be attached to one catalogue record. So in order to match a PDF to metadata, you need to visit all the individual ULIB and ELIB detail record views of the catalogue and extract the attached PDF file names from the HTML pages.

Since that is a crazy amount of work and the entire NBUV catalogue is archived in the Wayback Machine, we have not bothered with extracting the metadata from the HTML pages. It's just one bridge to far for us, even though we have a special IRBIS taskforce, which only did custom scraping scripts for IRBIS catalogues.

TL;DR: the NBUV catalogue is not offline, just unstable; all the PDFs are safe; all the metadata is preserved in the HTML pages captured by the Wayback Machine. So there's no need to crawl the NBUV catalogue again, we would probably just add to the pressure on the servers. And as long as the NBUV catalogue website is still up I would not recommend ingesting the PDFs into the IA yet, unless the National Librarian of Ukraine asks us to. But as you rightly pointed out NBUV has much more pressing things to worry about now.

**Comment by LeadSongDog** [comment]
We’ll, if the data is all captured, we can take a breath and think how best to proceed. Possibly  a blessing in disguise: OpenLibrary is fairly flexible in adapting to other cataloguing systems, even if less rigorous than others. It would probably be sensible to start with creating author records, giving both Roman and Cyrillic spellings in the a.k.a.

**Comment by bicolino34** [comment]
@cdrini should I still contact them or wait for now? I found this page with listed ways to contact them http://nbuv.gov.ua/node/18

**Comment by ekansa** [comment]
Hi All,

I'm happy to (try) to help if I can. I lack Ukranian language expertise or specific expertise in library metadata however. But if there's something that I can do with a bit of Python to scrape through some archived HTML that may contain useful metadata, please let me know!

**Comment by cdrini** [comment]
Thank you so much @storytracer !!! 😊 That's fantastic to hear! And thank you so much for the work you're doing with SUCHO!

There are a few points of integrations:

1. Copy the _catalogue_ into Open Library. These would make all the books appear in search results, but they would not appear read-able on Open Library. The catalogue would require coercing into Open Library's format, which seems like it would be a decent amount of work!
2. Have "Read" buttons that link to irbis-nbuv.gov.ua on Open Library . This is what our "Trusted Book Providers" program does; it provides a way to drive traffic to trusted providers while giving Open Library patrons access to more books. The way we generally do this is by storing an identifier on the the Open Library edition record, which we can use to create the "Read" link. It sounds like getting stable NBUV identifiers would be difficult! We can also make exceptions to this flow.
3. Back-up the PDFs on archive.org . This would require communication with NBUV, but would give the books a reliable home/backup. They could then also be viewed in the browser using archive.org 's BookReader without needing to download the PDFs, and even if NBUV is down. NBUV could also link to archive.org to reduce its network strain if it's useful. But this would require discussion with Alexis at the collections team on archive.org as well.

@ekansa would you be interested in helping with (1) above? In a nutshell, we'd need to crawl and convert all the HTML records to JSON objects of this shape: https://github.com/internetarchive/openlibrary-client/blob/master/olclient/schemata/import.schema.json , and then we can bulk import all those JSON records in Open Library.

**Comment by cdrini** [comment]
@bicolino34 Let me ping Alexis on our end here first to get some more info before we contact them! Though heads up might be a week or so before I have the time :)
